pkg/clients/aws.go

```diff
@@ -97,39 +96,48 @@ func UseProviderConfig(ctx context.Context, c client.Client, mg resource.Managed
 	switch s := pc.Spec.Credentials.Source; s { //nolint:exhaustive
 	case xpv1.CredentialsSourceInjectedIdentity:
 		cfg, err := UsePodServiceAccount(ctx, []byte{}, DefaultSection, region)
-		return SetResolver(ctx, mg, cfg), err
+		if err != nil {
+			return nil, err
+		}
+		return SetResolver(pc, cfg), nil
 	default:
 		data, err := resource.CommonCredentialExtractor(ctx, s, c, pc.Spec.Credentials.CommonCredentialSelectors)
 		if err != nil {
 			return nil, errors.Wrap(err, "cannot get credentials")
 		}
 		cfg, err := UseProviderSecret(ctx, data, DefaultSection, region)
-		return SetResolver(ctx, mg, cfg), err
+		if err != nil {
+			return nil, err
+		}
+		return SetResolver(pc, cfg), nil
 	}
 }
 
 // SetResolver parses annotations from the managed resource
 // and returns a configuration accordingly.
-func SetResolver(ctx context.Context, mg resource.Managed, cfg *aws.Config) *aws.Config {
-	if ServiceID, ok := mg.GetAnnotations()["aws.alpha.crossplane.io/endpointServiceID"]; ok {
-		if URL, ok := mg.GetAnnotations()["aws.alpha.crossplane.io/endpointURL"]; ok {
-			endpoint := aws.Endpoint{
-				URL: URL,
-			}
-			if Region, ok := mg.GetAnnotations()["aws.alpha.crossplane.io/endpointSigningRegion"]; ok {
-				endpoint.SigningRegion = Region
-			}
-
-			endpointResolver := func(service, region string) (aws.Endpoint, error) {
-				if strings.Contains(ServiceID, service) {
-					return endpoint, nil
-				}
-
-				return endpoint, &aws.EndpointNotFoundError{}
+func SetResolver(pc *v1beta1.ProviderConfig, cfg *aws.Config) *aws.Config {
+	if pc.Spec.Endpoint == nil {
+		return cfg
+	}
+	cfg.EndpointResolver = aws.EndpointResolverFunc(func(service, region string) (aws.Endpoint, error) {
+		e := aws.Endpoint{
+			URL:               pc.Spec.Endpoint.URL,
```

Comment:
I went back to initial [implementation PR](https://github.com/crossplane/provider-aws/pull/382#discussion_r504922347) and it seems like we need to construct the URL using `service` and `region` here.

Reply:
@hasheddan @chlunde Alright, I have changed the struct a little bit to account for dynamic and static URL resolving, which should cover all cases SDK allows. Tested the static config with localstack and dynamic config with `amazonaws.com`, added examples for both. It should be ready for another review.